app/src/main/java/pm/gnosis/heimdall/data/repositories/impls/PlayBillingRepository.kt

```diff
@@ -0,0 +1,164 @@
+package pm.gnosis.heimdall.data.repositories.impls
+
+import android.app.Activity
+import android.content.Context
+import com.android.billingclient.api.*
+import com.gojuno.koptional.Optional
+import com.gojuno.koptional.toOptional
+import io.reactivex.Single
+import io.reactivex.SingleEmitter
+import io.reactivex.SingleOnSubscribe
+import io.reactivex.rxkotlin.subscribeBy
+import pm.gnosis.heimdall.data.repositories.BillingRepository
+import pm.gnosis.heimdall.data.repositories.models.UserPurchase
+import pm.gnosis.heimdall.helpers.SetStore
+import pm.gnosis.svalinn.common.di.ApplicationContext
+import timber.log.Timber
+import java.lang.ref.WeakReference
+import java.util.*
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class PlayBillingRepository @Inject constructor(
+    @ApplicationContext private val context: Context
+) : BillingRepository, PurchasesUpdatedListener {
+
+    private val purchaseStore = SetStore<String>()
+
+    private val billingClient = BillingClient.newBuilder(context).setListener(this).build()
+
+    private val pendingActionsLock = Any()
+    private val pendingActions = ArrayList<BillingAction>()
+
+    private var isConnected = false
+
+    override fun init() {
+        updatePurchases()
```

Comment:
I would keep track of this disposable in `HeimdallApplication` since it is tied to that lifecycle.